### ISSUE TITLE
Logic to render a list of bucket-list items

### ISSUE THREAD

**Issue by djbowen95** [djbowen95/Hit-the-Bucket#5]
This is the fifth criteria found on the [brief](https://github.com/djbowen95/Hit-the-Bucket/blob/main/docs/BRIEF.md):
`It's done when I write logic to render a list of bucket-list items using .map.`

However, while it is listed fifth, I think it would be worth doing earlier - as once the application is rendering the bucket list / `Bucket` component correctly, it will easy to test changes and see if my solutions to the other tasks from the brief are working.

The tasks involved in resolving this issue are:
- [x] Render the `Bucket` component without breaking errors;
- [x] Pass the `bucket` state object from `BucketList` to `Bucket`;
- [x] Use `.map` to iterate through the array of objects stored in `bucket`;
- [x] Use properties from each object to built a JSX element;
- [x] Use properties to give every element a unique key - [as this is a requirement for React](https://sentry.io/answers/understanding-unique-keys-for-array-children-in-react-js/#:~:text=When%20you%20render%20an%20array,determine%20how%20elements%20have%20changed.)
- [x] Manually test by adding new bucket list items through the bucket list form

**Comment by djbowen95** [comment]
# Simple display function
While working on issue #1, I rendered the values stored by the `bucket` state object inside a simple `<h1>` element, in the main `BucketList` component. This made it easier to know if my changes to state were working a intended; it later helped build a skeleton for **rendering the bucket list items stored in state** in the `Bucket` component. 

- I started with most simple rendering: a **JSX expression** that took the state object, at the time **an array of string values**, and rendered it as part of the `BucketList` component. By default, it returned the values as a concatenated string.
 `<h1>Bucket: {bucket}</h1>`

- When I **changed the data type** of the values in the array - from strings to objects - the values would no longer render to the page. The application provided me with the following error:
 `Objects are not valid as a React child (found: object with keys {id, text, eagerness}). If you meant to render a collection of children, use an array instead.`

- I resolved this with a function which would **iterate** through the `bucket` items with `.map` and return a **JSX element**: a `<li>` item with a **unique key** ([as required by React](https://sentry.io/answers/understanding-unique-keys-for-array-children-in-react-js/#:~:text=When%20you%20render%20an%20array,determine%20how%20elements%20have%20changed.)), and displayed the bucket list item as inner text content.
```
const displayBucket = () => {
    const listItems = bucket.map((item) => {
      return (
        <li className="bucket-list-item" key={item.id}>
          {item.text}
        </li>
      );
    });
    return listItems;
```

- I used a **JSX expression** to run this function; containing it in an `ul` element:
`<ul>{displayBucket()}</ul>`

While ultimately not part of the final product, this method of rendering the bucket list items was good enough while I was developing the logic to add a bucket list item (#1).

**Comment by djbowen95** [comment]
# Integrating with the Bucket component
The starter code came with an existing frame for rendering the bucket list: the `Bucket` component. Once I had completed the logic to add a bucket list item (#1), I worked to make the changes required to have this component render elements based on the values stored in the `bucket` state object. To do this, I:

- Imported the `Bucket` component as a module in the `BucketList` component. This was initially a breaking change: as **calls to functions that did not exist yet** and **elements assigned empty JSX elements as attributes** prevented `Bucket` from rendering.
`import Bucket from './Bucket';`

- Deleted or commented out every function call or empty attribute from the component's return value:
` <p onClick={ }> ✏️</p> ...became... <p>✏️</p>`

- Passed the `bucket` state object through from the `BucketList` component to the `Bucket` component **as props**; as I needed to read its values to render them in the `Bucket` component. 

- Iterated through the items in the `bucket` array; rendered each element with text content and a key **drawn from the properties of the objects** in the array.
 `<div key={item.id}>`
 `     {item.text}`
 `  </div>`

**Comment by djbowen95** [comment]
The existing `Bucket` component returned each item in a container `div` with several class names. There was pre-built styling for these elements in the starter code, to demonstrate whether or not a bucket list item was complete, and different levels of 'eagerness' for a particular item. 

Writing code that would change this formatting was the goal of issue #6: but as applying these styles also required reading and interpretting properties from the `bucket` item, I solved this as I built the logic to render the text content of the bucket list items in `Bucket` component. I've outlined what I did to solve this under that issue.
